This fixes a crash on exit that I get repeatidly:
`[ERROR] [socket_can_receiver_node_exe-1]: process has died [pid 10882, exit code -6, cmd '/home/user/ros2/install/ros2_socketcan/lib/ros2_socketcan/socket_can_receiver_node_exe --ros-args -r __node:=socket_can_receiver -r __ns:=/ --params-file /tmp/launch_params_1176h11h -r from_can_bus:=/can_bus_dbw/can_rx -r from_can_bus_fd:=/can_bus_dbw/can_rx'].`

--- a/ros2_socketcan/include/ros2_socketcan/socket_can_receiver_node.hpp
+++ b/ros2_socketcan/include/ros2_socketcan/socket_can_receiver_node.hpp
@@ -17,6 +17,7 @@
 #ifndef ROS2_SOCKETCAN__SOCKET_CAN_RECEIVER_NODE_HPP_
 #define ROS2_SOCKETCAN__SOCKET_CAN_RECEIVER_NODE_HPP_
 
+#include <atomic>
 #include <memory>
 #include <thread>
 #include <string>
@@ -49,9 +50,9 @@ public:
   /// \brief Default constructor
   explicit SocketCanReceiverNode(rclcpp::NodeOptions options);
 
-  /// \brief Destructor. Joins the receiver thread if it is still joinable, since destroying a
-  /// joinable std::thread calls std::terminate(). This happens on plain shutdown (e.g. SIGINT),
-  /// where the lifecycle node is destroyed without transitioning through "cleanup".
+  /// \brief Destructor. Stops and joins the receiver thread, since destroying a joinable
+  /// std::thread calls std::terminate(). Needed on plain shutdown (e.g. SIGINT), where the
+  /// lifecycle node is destroyed without transitioning through "cleanup".
   ~SocketCanReceiverNode() override;
 
   /// \brief Callback from transition to "configuring" state.
@@ -78,11 +79,15 @@ public:
   void receive();
 
 private:
+  /// \brief Stop the receiver thread and join it if it exists.
+  void stop_receiver_thread();
+
   std::string interface_;
   std::shared_ptr<lc::LifecyclePublisher<can_msgs::msg::Frame>> frames_pub_;
   std::shared_ptr<lc::LifecyclePublisher<ros2_socketcan_msgs::msg::FdFrame>> fd_frames_pub_;
   std::unique_ptr<SocketCanReceiver> receiver_;
   std::unique_ptr<std::thread> receiver_thread_;
+  std::atomic<bool> stop_thread_{false};
   std::chrono::nanoseconds interval_ns_;
   bool enable_fd_;
   bool use_bus_time_;

--- a/ros2_socketcan/include/ros2_socketcan/socket_can_receiver_node.hpp
+++ b/ros2_socketcan/include/ros2_socketcan/socket_can_receiver_node.hpp
@@ -49,6 +49,11 @@ public:
   /// \brief Default constructor
   explicit SocketCanReceiverNode(rclcpp::NodeOptions options);
 
+  /// \brief Destructor. Joins the receiver thread if it is still joinable, since destroying a
+  /// joinable std::thread calls std::terminate(). This happens on plain shutdown (e.g. SIGINT),
+  /// where the lifecycle node is destroyed without transitioning through "cleanup".
+  ~SocketCanReceiverNode() override;
+
   /// \brief Callback from transition to "configuring" state.
   /// \param[in] state The current state that the node is in.
   LNI::CallbackReturn on_configure(const lc::State & state) override;

--- a/ros2_socketcan/src/socket_can_receiver_node.cpp
+++ b/ros2_socketcan/src/socket_can_receiver_node.cpp
@@ -67,9 +67,16 @@ SocketCanReceiverNode::SocketCanReceiverNode(rclcpp::NodeOptions options)
 
 SocketCanReceiverNode::~SocketCanReceiverNode()
 {
+  stop_receiver_thread();
+}
+
+void SocketCanReceiverNode::stop_receiver_thread()
+{
+  stop_thread_ = true;
   if (receiver_thread_ && receiver_thread_->joinable()) {
     receiver_thread_->join();
   }
+  receiver_thread_.reset();
 }
 
 LNI::CallbackReturn SocketCanReceiverNode::on_configure(const lc::State & state)
@@ -97,6 +104,8 @@ LNI::CallbackReturn SocketCanReceiverNode::on_configure(const lc::State & state)
     fd_frames_pub_ =
       this->create_publisher<ros2_socketcan_msgs::msg::FdFrame>("from_can_bus_fd", 500);
   }
+
+  stop_thread_ = false;
 
 #ifdef USE_AGNOCAST_ENABLED
   const std::string thread_name = "socket_can_receiver:" + interface_ + ":receiver_thread";
@@ -143,15 +152,14 @@ LNI::CallbackReturn SocketCanReceiverNode::on_cleanup(const lc::State & state)
 {
   (void)state;
 
+  stop_receiver_thread();
+
   if (!enable_fd_) {
     frames_pub_.reset();
   } else {
     fd_frames_pub_.reset();
   }
 
-  if (receiver_thread_ && receiver_thread_->joinable()) {
-    receiver_thread_->join();
-  }
   RCLCPP_DEBUG(this->get_logger(), "Receiver cleaned up.");
   return LNI::CallbackReturn::SUCCESS;
 }
@@ -159,6 +167,9 @@ LNI::CallbackReturn SocketCanReceiverNode::on_cleanup(const lc::State & state)
 LNI::CallbackReturn SocketCanReceiverNode::on_shutdown(const lc::State & state)
 {
   (void)state;
+
+  stop_receiver_thread();
+
   RCLCPP_DEBUG(this->get_logger(), "Receiver shutting down.");
   return LNI::CallbackReturn::SUCCESS;
 }
@@ -171,7 +182,7 @@ void SocketCanReceiverNode::receive()
     can_msgs::msg::Frame frame_msg(rosidl_runtime_cpp::MessageInitialization::ZERO);
     frame_msg.header.frame_id = "can";
 
-    while (rclcpp::ok()) {
+    while (rclcpp::ok() && !stop_thread_) {
       if (this->get_current_state().id() != State::PRIMARY_STATE_ACTIVE) {
         std::this_thread::sleep_for(100ms);
         continue;
@@ -205,7 +216,7 @@ void SocketCanReceiverNode::receive()
     ros2_socketcan_msgs::msg::FdFrame fd_frame_msg(rosidl_runtime_cpp::MessageInitialization::ZERO);
     fd_frame_msg.header.frame_id = "can";
 
-    while (rclcpp::ok()) {
+    while (rclcpp::ok() && !stop_thread_) {
       if (this->get_current_state().id() != State::PRIMARY_STATE_ACTIVE) {
         std::this_thread::sleep_for(100ms);
         continue;

--- a/ros2_socketcan/src/socket_can_receiver_node.cpp
+++ b/ros2_socketcan/src/socket_can_receiver_node.cpp
@@ -65,6 +65,13 @@ SocketCanReceiverNode::SocketCanReceiverNode(rclcpp::NodeOptions options)
   RCLCPP_INFO(this->get_logger(), "interval(s): %f", interval_sec);
 }
 
+SocketCanReceiverNode::~SocketCanReceiverNode()
+{
+  if (receiver_thread_ && receiver_thread_->joinable()) {
+    receiver_thread_->join();
+  }
+}
+
 LNI::CallbackReturn SocketCanReceiverNode::on_configure(const lc::State & state)
 {
   (void)state;
@@ -142,7 +149,7 @@ LNI::CallbackReturn SocketCanReceiverNode::on_cleanup(const lc::State & state)
     fd_frames_pub_.reset();
   }
 
-  if (receiver_thread_->joinable()) {
+  if (receiver_thread_ && receiver_thread_->joinable()) {
     receiver_thread_->join();
   }
   RCLCPP_DEBUG(this->get_logger(), "Receiver cleaned up.");

